viewer/src/public/migration/Cognite3DModel.ts

```diff
@@ -209,48 +221,72 @@ export class Cognite3DModel extends THREE.Object3D implements CogniteModelBase {
 
   async setNodeColor(nodeId: number, r: number, g: number, b: number): Promise<void> {
     const treeIndex = await this.nodeIdAndTreeIndexMaps.getTreeIndex(nodeId);
-    this.setNodeColorByTreeIndex(treeIndex, r, g, b);
+    await this.setNodeColorByTreeIndex(treeIndex, r, g, b);
   }
 
-  setNodeColorByTreeIndex(treeIndex: number, r: number, g: number, b: number) {
-    this.nodeColors.set(treeIndex, [r, g, b]);
-    this.cadNode.requestNodeUpdate([treeIndex]);
+  /**
+   * Update color of a 3D node identified by it's tree index.
+   * @param treeIndex       Tree index of the node to update
+   * @param r               Red component (0-255)
+   * @param g               Green component (0-255)
+   * @param b               Blue component (0-255)
+   * @param applyToChildren When true, the color will be applied to all descendants
+   * @returns               Promise that resolves to number of nodes affected
+   */
+  async setNodeColorByTreeIndex(
+    treeIndex: number,
+    r: number,
+    g: number,
+    b: number,
+    applyToChildren = false
+  ): Promise<number> {
+    const treeIndices = await this.determineTreeIndices(treeIndex, applyToChildren);
+    const color: [number, number, number] = [r, g, b];
+    treeIndices.forEach(idx => this.nodeColors.set(idx, color));
+    this.cadNode.requestNodeUpdate(treeIndices);
+    return treeIndices.count;
   }
 
   async resetNodeColor(nodeId: number): Promise<void> {
     const treeIndex = await this.nodeIdAndTreeIndexMaps.getTreeIndex(nodeId);
-    this.resetNodeColorByTreeIndex(treeIndex);
+    await this.resetNodeColorByTreeIndex(treeIndex);
   }
 
-  resetNodeColorByTreeIndex(treeIndex: number) {
-    this.nodeColors.delete(treeIndex);
-    this.cadNode.requestNodeUpdate([treeIndex]);
+  async resetNodeColorByTreeIndex(treeIndex: number, applyToChildren = false): Promise<number> {
```

Comment:
Same comment applies. I feel like people will do something like.
```
let selected: [] = [];

const select = (model: Cognite3DModel, treeIndices: []) => {
  if (selected.length > 0 ) {
    await model.resetNodeColorByTreeIndex(selected);
  }
  selected = treeIndices;
  await model.setNodeColorByTreeIndex(selected, selectedColor);
};
```

And wonder why it doesn't work.

Reply:
Hm, why won't that work?

Reply:
sorry, I was missing a line there..

```
let selected: [] = [];

const select = (model: Cognite3DModel, treeIndices: []) => {
  if (selected.length > 0 ) {
    model.deselectAllNodes();
    await model.resetNodeColorByTreeIndex(selected);
  }
  selected = treeIndices;
  await model.setNodeColorByTreeIndex(selected, selectedColor);
  treeIndices.forEach(treeIndex => model.selectNodeByTreeIndex(treeIndex));
};
```

If I am not mistaken the setColor sets what is returned by the model appearance when it isn't selected?

Reply:
The code you provide will cause the `selectedColor` to not have any effect untill `deselectNode` is called, yes.